avatica/server/src/main/java/org/apache/calcite/avatica/jdbc/JdbcMeta.java

```diff
@@ -852,6 +855,74 @@ public Frame fetch(StatementHandle h, long offset, int fetchMaxRowCount) throws
     }
   }
 
+  @Override public ExecuteBatchResult prepareAndExecuteBatch(StatementHandle h,
+      List<String> sqlCommands) throws NoSuchStatementException {
+    try {
+      // Get the statement
+      final StatementInfo info = statementCache.getIfPresent(h.id);
+      if (info == null) {
+        throw new NoSuchStatementException(h);
+      }
+
+      // addBatch() for each sql command
+      final Statement stmt = info.statement;
+      for (String sqlCommand : sqlCommands) {
+        stmt.addBatch(sqlCommand);
+      }
+
+      // Execute the batch and return the results
+      return new ExecuteBatchResult(stmt.executeBatch());
+    } catch (SQLException e) {
+      throw propagate(e);
+    }
+  }
+
+  @Override public ExecuteBatchResult executeBatch(StatementHandle h,
+      List<List<TypedValue>> updateBatches) throws NoSuchStatementException {
+    try {
+      final StatementInfo info = statementCache.getIfPresent(h.id);
+      if (null == info) {
+        throw new NoSuchStatementException(h);
+      }
+
+      final PreparedStatement preparedStmt = (PreparedStatement) info.statement;
+      for (List<TypedValue> batch : updateBatches) {
+        int i = 1;
+        for (TypedValue value : batch) {
+          // Use the value and then increment
+          preparedStmt.setObject(i++, value.value);
+        }
+        preparedStmt.addBatch();
+      }
+      return new ExecuteBatchResult(preparedStmt.executeBatch());
+    } catch (SQLException e) {
+      throw propagate(e);
```

Comment:
It could theoretically make sense to add "row index of the problematic statement" to the error message. In other words, something like "unable to setObject for row X, column Y"